app/models/feed.rb

```diff
@@ -7,15 +7,28 @@ def initialize(type, account)
   end
 
   def get(limit, max_id = nil, since_id = nil)
+    if redis.exists("account:#{@account.id}:regeneration")
+      from_database(limit, max_id, since_id)
+    else
+      from_redis(limit, max_id, since_id)
+    end
+  end
+
+  private
+
+  def from_redis(limit, max_id, since_id)
     max_id     = '+inf' if max_id.blank?
     since_id   = '-inf' if since_id.blank?
     unhydrated = redis.zrevrangebyscore(key, "(#{max_id}", "(#{since_id}", limit: [0, limit], with_scores: true).map(&:last).map(&:to_i)
     status_map = Status.where(id: unhydrated).cache_ids.map { |s| [s.id, s] }.to_h
-
     unhydrated.map { |id| status_map[id] }.compact
   end
 
-  private
+  def from_database(limit, max_id, since_id)
+    results = Status.as_home_timeline(@account).paginate_by_max_id(limit, max_id, since_id).to_a
```

Comment:
@akihikodaki I wanted to do it this way to avoid copying the array (reject! mutates existing array afaik)

@clworld Yes, that's a problem. Maybe limit should be doubled? Or do second request if the result of reject! is empty.. But even then there is no guarantee new items won't be rejected.

I am reminded that @akihikodaki wrote a version of the home query that includes the rules from FeedManager#filter?. Perhaps this is the place to dig that up and use it...